tools/purge_executions.py

```diff
@@ -58,101 +56,110 @@ def do_register_cli_opts(opts, ignore_errors=False):
+                    ', "failed", "canceled" and "timed_out" are deleted.'),
+    ]
+    _do_register_cli_opts(cli_opts)
+
+
+def _purge_models(execution_db):
+    liveaction_id = execution_db.liveaction.get('id', None)
 
     if not liveaction_id:
-        print('Invalid LiveAction id. Skipping delete: %s', execution_db)
+        LOG.error('Invalid LiveAction id. Skipping delete: %s', execution_db)
 
     liveaction_db = None
     try:
         liveaction_db = LiveAction.get_by_id(liveaction_id)
     except:
-        print('LiveAction with id: %s not found. Skipping delete.', liveaction_id)
+        LOG.exception('LiveAction with id: %s not found. Skipping delete.', liveaction_id)
     else:
         global DELETED_COUNT
         DELETED_COUNT += 1
 
     try:
         ActionExecution.delete(execution_db)
-    except Exception as e:
-        print('Exception deleting Execution model: %s, exception: %s',
-              execution_db, str(e))
+    except:
+        LOG.exception('Exception deleting Execution model: %s',
+                      execution_db)
     else:
-        try:
-            LiveAction.delete(liveaction_db)
-        except Exception as e:
-            print('Zombie LiveAction left in db: %s. Exception: %s',
-                  liveaction_db, str(e))
+        if liveaction_db:
+            try:
+                LiveAction.delete(liveaction_db)
+            except:
+                LOG.exception('Zombie LiveAction left in db: %s.', liveaction_db)
+
 
+def _should_delete(execution_db, action_ref, timestamp, purge_incomplete=False):
+    if not purge_incomplete and execution_db.status in IN_PROGRESS_STATES:
+        return False
 
-def _purge_executions(timestamp=None, action_ref=None):
+    if action_ref != '':
+        return (execution_db.liveaction['action'] == action_ref and
+                execution_db.start_timestamp < timestamp)
+    else:
+        return execution_db.start_timestamp < timestamp
+
+
+def purge_executions(timestamp=None, action_ref=None, purge_incomplete=False):
     if not timestamp:
-        print('Specify a valid timestamp to purge.')
+        LOG.error('Specify a valid timestamp to purge.')
         return
 
     if not action_ref:
         action_ref = ''
 
-    print('Purging executions older than timestamp: %s' %
-          timestamp.strftime('%Y-%m-%dT%H:%M:%S.%fZ'))
-
-    def should_delete(execution_db):
-        if action_ref != '':
-            return (execution_db.liveaction['action'] == action_ref and
-                    execution_db.start_timestamp < timestamp)
-        else:
-            return execution_db.start_timestamp < timestamp
+    LOG.info('Purging executions older than timestamp: %s' %
+             timestamp.strftime('%Y-%m-%dT%H:%M:%S.%fZ'))
 
     # XXX: Think about paginating this call.
-    filters = {'start_timestamp__lt': isotime.parse(timestamp)}
+    filters = {'end_timestamp__lt': isotime.parse(timestamp)}
     executions = ActionExecution.query(**filters)
```

Comment:
Likely a richer filter i.e. taking into account 'action-ref' and 'purge-incomplete' would be better.

Wonder if there is a way to only retrieve ids instead of full objects.

Btw, I think we can even leverage https://github.com/StackStorm/st2/blob/master/st2common/st2common/models/db/__init__.py#L183. This would avoid unnecessary reads.